### ISSUE TITLE
Issue with uri path variable containing ?

### ISSUE THREAD

**Issue by tootoohu** [http4k/http4k#623]
I defined an API like this - GET /member/{text}, {text} is the path variable. I got issues with some special chars containing in {text}, for example:
1. if this value is passed containing ? like "?861246", then the request can't be routed correctly,  my controller with defining GET /member/{text} can't be found, 404 will be returned back.
2. If this value is passed like "877?123", the chars after ? will be truncated, {text} only get the value of "877", that is what I want.

All I want to get is the raw data from client.  Thank you~
version: 3.112.2

**Comment by daviddenton-zuhlke** [comment]
Question marks in URL delimit the main PATH of the URL from the QUERY.

```kotlin
fun main() {
    val uri = Uri.of("http://localhost/abd?cde=fgh")
    println(uri)
    println(uri.path)
    println(uri.query("cde"))
}
```

I'll not go into it here as http4k is acting as intended, but you can get more information at this link:

https://doepud.co.uk/blog/anatomy-of-a-url

**Comment by daviddenton** [comment]
Question marks in URL delimit the main PATH of the URL from the QUERY.

```kotlin
fun main() {
    val uri = Uri.of("http://localhost/abd?cde=fgh")
    println(uri)
    println(uri.path)
    println(uri.query("cde"))
}
```

I'll not go into it here as http4k is acting as intended, but you can get more information at this link:

https://doepud.co.uk/blog/anatomy-of-a-url

**Comment by npryce** [comment]
@tootoohu the client needs to escape characters that have special meaning in URL syntax if you want to pass them in URL paths.  For example, to send "877?123", you need to escape it like: http://localhost/member/877%3F123

The server will unescape the path element.  When you get the `text` parameter from the Request in your HttpHandler, HTTP4K will unescape it and return it as "877?123".

**Comment by tootoohu** [comment]
> @tootoohu the client needs to escape characters that have special meaning in URL syntax if you want to pass them in URL paths. For example, to send "877?123", you need to escape it like: http://localhost/member/877%3F123
> 
> The server will unescape the path element. When you get the `text` parameter from the Request in your HttpHandler, HTTP4K will unescape it and return it as "877?123".

Thank you @npryce 
I encoded the string "877?123" in the request to  members/877%3F123, I debugged to  my controller, the value of request.uri.path is "877"

**Comment by daviddenton** [comment]
I'm not sure what is going on with you use-case @tootoohu - this runnable code shows that the library is working as expected. 😄 
```kotin
fun main() {
    { req: Request ->
        Response(OK).body(req.uri.path)
    }.asServer(SunHttp(9000)).start()

    val uri = Uri.of("http://localhost:9000/members/877%3F123")

    val client = JavaHttpClient()

    println(client(Request(GET, uri)).bodyString()).  // prints /members/877%3F123
}
```

You should also update to the latest version of http4k to keep on top of all issues which we have previously fixed we are currently at 4.9.1.0. If you can't go to v4, then the latest version ov v3 was 3.285.2

**Comment by tootoohu** [comment]
> I'm not sure what is going on with you use-case @tootoohu - this runnable code shows that the library is working as expected. 😄
> 
> ```
> fun main() {
>     { req: Request ->
>         Response(OK).body(req.uri.path)
>     }.asServer(SunHttp(9000)).start()
> 
>     val uri = Uri.of("http://localhost:9000/members/877%3F123")
> 
>     val client = JavaHttpClient()
> 
>     println(client(Request(GET, uri)).bodyString()).  // prints /members/877%3F123
> }
> ```
> 
> You should also update to the latest version of http4k to keep on top of all issues which we have previously fixed we are currently at 4.9.1.0. If you can't go to v4, then the latest version ov v3 was 3.285.2

thank you
well, I think I  found the root cause, why there is some strange result with the URI path, that is because the undertow is used which also decodes the path by default.

**Comment by tootoohu** [comment]
I found the cause and mentioned above, thank you ALL very much for the investigation and quick response~
